### PR TITLE
Account for associated type, const and static when using type argument

### LINKED ISSUE #45447: Bad error message using associated type
I think this code shows a bad error message:

```
trait Foo { const FOO: Self; }
impl Foo for u32 { const FOO: Self = 1; }
fn bar<T: Foo>(n: T) {
    const BASE: T = T::FOO;
}
fn main() {}
```

rustc 1.22.0-nightly (4279e2b4c 2017-10-21) gives:

```
error[E0401]: can't use type parameters from outer function; try using a local type parameter instead
 --> ...\test.rs:4:17
  |
4 |     const BASE: T = T::FOO;
  |                 ^ use of type variable from outer function

error[E0401]: can't use type parameters from outer function; try using a local type parameter instead
 --> ...\test.rs:4:21
  |
4 |     const BASE: T = T::FOO;
  |                     ^^^^^^ use of type variable from outer function

error: aborting due to 2 previous errors
```

### PR DESCRIPTION
**Pull request by estebank**
When using a type argument from the enclosing scope in an associated type,
const or static item provide a more targetted E0401 error, instead of
blaming the "outer function" blindly.

Fix #45447.

### CODE REVIEW

File: src/librustc_resolve/lib.rs [hunk 1]
```diff
@@ -1691,6 +1745,9 @@ pub struct Resolver<'a> {
     current_type_ascription: Vec<Span>,
 
     injected_crate: Option<Module<'a>>,
+
+    /// Used for more accurate error when using type parameters in a associated items.
+    scope: CurrentScope,
```
Comment (petrochenkov) [inline src/librustc_resolve/lib.rs]: Please use some other terminology than "scope", it's incredibly overloaded in resolve.
Also, could you move `enum CurrentScope` and its impls into `diagnostics.rs`?

File: src/test/ui/inner-static-type-parameter.rs [hunk 2]
```diff
@@ -4,7 +4,7 @@ enum Bar<T> { What } //~ ERROR parameter `T` is never used
 
 fn foo<T>() {
     static a: Bar<T> = Bar::What;
-//~^ ERROR can't use generic parameters from outer function
+    //~^ ERROR can't use generic parameters in `static` associated item
```
Comment (Centril) [inline src/test/ui/inner-static-type-parameter.rs]: This `static` is not an associated item (and cannot be as that would entail generic statics!). It is a free item belonging to `fn foo` as a "module".

File: src/test/ui/issues/issue-45447.rs [hunk 3]
```diff
@@ -0,0 +1,10 @@
+trait Foo { const FOO: Self; }
+impl Foo for u32 { const FOO: Self = 1; }
+fn bar<T: Foo>(n: T) {
+    const BASE: T = T::FOO;
+    //~^ ERROR can't use generic parameters in `const` associated item
```
Comment (Centril) [inline src/test/ui/issues/issue-45447.rs]: This is also wrong; not an associated item.

File: src/librustc_resolve/lib.rs (hunk 1 above)
Comment (estebank) [reply src/librustc_resolve/lib.rs]: What would be a better name?

Comment (cramertj) [reply src/librustc_resolve/lib.rs]: `BodyOwner` / `DefOwner` maybe? (though `Def` is similarly unfortunately overloaded)

File: src/test/ui/issues/issue-45447.rs [hunk 4]
```diff
@@ -0,0 +1,10 @@
+trait Foo { const FOO: Self; }
+impl Foo for u32 { const FOO: Self = 1; }
+fn bar<T: Foo>(n: T) {
+    const BASE: T = T::FOO;
+    //~^ ERROR can't use generic parameters in a `const`
+    //~| ERROR can't use generic parameters in a `const`
+    type Type = T;
+    //~^ ERROR can't use generic parameters in associated type
```
Comment (Centril) [inline src/test/ui/issues/issue-45447.rs]: Here `Type` is not an associated type alias but rather a free type alias.

File: src/librustc_resolve/lib.rs [hunk 5]
```diff
@@ -258,7 +273,14 @@ fn resolve_struct_error<'sess, 'a>(resolver: &'sess Resolver<'_>,
             // Try to retrieve the span of the function signature and generate a new message with
             // a local type or const parameter.
             let sugg_msg = &format!("try using a local generic parameter instead");
-            if let Some((sugg_span, new_snippet)) = cm.generate_local_type_param_snippet(span) {
+            if !resolver.scope.is_other() {
+                err.help(&format!(
+                    "{} need a type instead of a generic parameter",
```
Comment (Centril) [inline src/librustc_resolve/lib.rs]:
```suggestion
                    "{} needs a type instead of a generic parameter",
```

**Comment by gagan0723** [comment]
ping from triage @estebank  any updates on this?

**Comment by gagan0723** [comment]
ping from triage @cramertj this PR is waiting for your review?

**Comment by estebank** [comment]
I have some WIP code in my local branch to address some of the comments and will need a rebase. I won't be getting to it until at least the weekend.

**Comment by estebank** [comment]
Closing, I'll reopen when I get back to this.

**Pull request closed without merging**